README.md
<h1 align="center">
SWARM
</h1>

<p><img src="https://raw.githubusercontent.com/tanglebay/swarm/master/logo/swarm_gh_banner.png?sanitize=true"></p>

<p align="center">
SWARM is a script that allows you to easily install and manage an IOTA node. The most important configuration options are given via simple and selectable menu items. About the additional feature "Watchdog", the status of the nodes can be monitored, as well as automatically install updates. A reverse proxy installation is also offered for secure access via HTTPS.
</p>

<p align="center">
    <a href="https://twitter.com/TANGLEBAY" style="text-decoration:none;"><img src="https://img.shields.io/badge/Twitter-9cf.svg?logo=twitter" alt="Twitter"></a>
    <a href="https://discord.tanglebay.com/" style="text-decoration:none;"><img src="https://img.shields.io/badge/Discord-9cf.svg?logo=discord" alt="Discord"></a>
</p>

---

## Tutorials
### By dumdave#5233
- <a href="https://tangledust.com/install_iota_hornet_node_with_SWARM.html">Installing an IOTA Hornet node on a Netcup virtual server using SWARM</a>
- <a href="https://tangledust.com/install_iota_bee_node_with_SWARM.html">Installing an IOTA Bee node on a Netcup virtual server using SWARM</a>
- <a href="https://tangledust.com/install_iota_goshimmer_node_with_SWARM.html">Installing an IOTA GoShimmer node on a Netcup virtual server using SWARM</a>
- <a href="https://tangledust.com/install_iota_wasp_node_with_SWARM.html">Installing an IOTA WASP node on a Netcup virtual server using SWARM</a>
### By PIOTA
- <a href="https://iota-industrie-4-0.blogspot.com/2021/10/tutorial-zur-installation-einer-iota.html">Tutorial zur Installation einer IOTA-Fullnode mit SWARM</a>

---

## Installation

### Requirements
- VPS (2Core/4GB/40GB) **(GoShimmer/Bee cannot be used on ARM)**
- Ubuntu 20.04 LTS (64Bit)
- An existing (own) domain

**Download the latest release version of SWARM and run it. Do the following steps:**

1. Install SWARM with the Installer
```shell
curl -sL https://raw.githubusercontent.com/tanglebay/swarm/master/installer.sh | sudo bash -
```
2. After installting SWARM you can execute it with typing `swarm`

**Notice**: Please make sure that your node is properly secured (e.g. firewall, SSH KeyAuth, regular system updates).

---

## Using SWARM
In the menu of SWARM you navigate with the arrow keys. You confirm an action with the "OK" item or by pressing the Enter key. The "Cancel" item cancels an action or returns to the previous menu.

The configuration of the node should always be done via the configuration point provided by SWARM. It is not advisable to use the advanced configuration, since direct parameters set via this menu item are not taken into account during an update and are therefore reset.

---

## Ports
### Hornet
- Gossip port (peering): 15600/tcp (customizable)
- Autopeering port (peering): 14626/udp (customizable)
- Prometheus: 9311/tcp (if the plugin is activated)
- Mqtt: 1833/tcp (if the plugin is activated)
### Bee
- Gossip port (peering): 15601/tcp (customizable)
### GoShimmer
- Gossip port (peering): 14666/tcp (customizable)
- Autopeering port (peering): 14646/udp
- Prometheus: 9313/tcp (if the plugin is activated)
### Wasp
- Peering port: 4000/tcp (customizable)
- Nanomsg port: 5550/tcp (customizable)
### Proxy
In the proxy menu you can define a separate domain for each node or one domain for all. Please note that when using a shared domain, a separate port must be defined for each node.
- HTTP: 80/tcp

**Please note that the following ports should not be used for the proxy 8080/tcp-8084/tcp**

---

## Backup SWARM
If you ever want to reinstall your server or SWARM, you can backup your SWARM configuration first.

You can find the configuration files under the following path:
```shell
cd /var/lib/swarm/configs
```

After the reinstallation, simply replace the configurations with your backup, run a SWARM update to update the configuration files (if necessary) and then you can reinstall Hornet and the proxy and have all settings imported directly.

---

## Disclaimer

**Please note that you use this script at your own risk and that I will not be liable for any damages that may occur**

---

## Donate

**IOTA Address:**
```
iota1qpzg5s4gmcp0khxlh0agt2cpspx0cmqr0dqjdutsn844565vj4gyjgt996w
```

CHANGELOG.md
## Version 1.0.5
### Added
- Added new Bee peers configuration
### Changed
- Updated proxy templates for Hornet/bee

---

## Version 1.0.3.2
### Changed
- Fixed proxy webapi status in waspInfo
- Disabled WASP installation

---

## Version 1.0.3.1
### Changed
- Fixed proxy deployent
- Fixed WASP admin whitelist configuration

---

## Version 1.0.3
- Hotfix proxy

---

## Version 1.0.2
- Hotfix for Bee parser
- Hotfix for Hornet parser

---

## Version 1.0.1
### Added
- Added option to allow/deny GoShimmer db download by Watchdog
- Added lock file for GoShimmer to prevent an error by Watchdog
- Added Bee autopeering (disabled until release)
- Added WASP tutorial by dumdave#5233 to readme
### Changed
- Updated parsing to used network config
- Updated proxy templates
- Updated proxy configuration
- Fixed CL editor output

---

## Version 1.0.0
### Added
- Added check for automatic proxy deploy (dev)
- Added option to select command-line editor in SWARM
- Added webAPI admin whitelist in WASP configuration
### Changed
- Fixed existing mainnetdb folder for goshimmer install process
- Fixed missing node removal from landing page
- Fixed missing default domain in proxy landing section
- Fixed goshimmer/wasp auth
- Fixed goshimmer proxy removal on goshimmer removal
- Fixed low check rate for goshimmer sync check
- Fixed GUI size
- Fixed proxy automatic deployment
- Fixed Hornet/Bee dashboard peers configuration
- Updated SWARM menu and splitted SWARM/Watchdog configuration
- Updated node information page
- Updated swarmUpdater to prevent errors
- Updated proxyInfo to hide urls of not installed nodes
- Updated node config parsing
### Removed
- Remove timeFrame sync check for GoShimmer

---

## Version 0.9.9
### Added
- Added Telegram Notifications for Watchdog
- Added WASP-CLI for WASP ID/peering
- Added alias for GoShimmer/WASP [Watchdog]
- Added root landing page
### Changed
- Updated waspInfo informations
- Fixed port parsing in the proxy
- Fixed auth for GoShimmer/WASP
- Fixed missing exitStatus
- Fixed waspAllowDbReset for WASP

---

## Version 0.9.8
### Changed
- Fixed WASP proxy configuration

---

## Version 0.9.7
### Added
- Added proxy deployment with SWARM update
- Added waspPeeringId to configuration
- Added own proxy settings for WASP API
### Changed
- Fixed duplicate error in wasp proxy template
- Fixed missing wasp check in proxyMenu
- Fixed wrong VAR for WASP dashboard auth
- Fixed GoShimmer dashboard/webapi auth
- Fixed missing _blank in index.html
- Updated proxy deployment

---

## Version 0.9.6
### Added
- Added peeringID configuration for WASP
### Changed
- Fixed Bee dashboard link
- Fixed Bee username configuration
- Fixed WASP dashboard password configuration
- Updated readme

---

## Version 0.9.5
### Added
- Added WASP to SWARM
- Added landing page for all nodes
- Added status monitor for all nodes to langdingpage
### Changed
- Fixed all missing exitStatus
- Fixed general bugs

---

## Version 0.9.4
### Added
- Added routes configuration in Hornet configuration
- Added CORS to proxy templates
### Changed
- Updated proxy template for GoShimmer
- Fixed sync check for GoShimmer
- Fixed GoShimmer configuration for the dashboard
### Removed
- Removed faucet from Hornet plugin menu
- Removed parser for snapshot DL in Bee
- Removed port checks for Hornet, Bee and GoShimmer

---

## Version 0.9.3
### Added
- Added prometheus for Hornet and Goshimmer
### Changed
- Fixed plugin management
- Fixed Goshimmer version module
- Fixed general bugs

---

## Version 0.9.2
### Added
- Added new plugin management for Hornet and GoShimmer
### Changed
- Fixed missing theme colors
- Fixed db download when performing auto update for GoShimmer
- Fixed GoShimmer enable plugin module
- Fixed Goshimmer database reset

---

## Version 0.9.1
### Changed
- Fixed Bee config check
### Removed
- Removed IF snapshot source from Bee

---

## Version 0.9.0
### Changed
- Moved Bee Coo public key to parser
- Moved Snapshot DL URL to Bee parser
- Updated Bee identity generation
- Updated Bee pwd generation
- Updated Coo public keys for Bee
- Update SWARM configuration
- Fixed Bee proxy template

---

## Version 0.8.9
### Added
- Added GoShimmer DB download
- Added Devnet to Bee
- Added Devnet to Hornet
### Changed
- Fixed parsing of Goshimmer autopeering port
- Fixed Hornet configuration menu
- Fixed general bugs

---

## Version 0.8.8
### Changed
- Updated goshimmerAutopeeringPort
- Updated goshimmerInstaller
- Updated GoShimmer DB reset

---

## Version 0.8.7
### Added
- Added OS architecture detection
- Added random update time frame for auto node update
- Added watchdog.cfg
### Changed
- Updated menus of Hornet, Bee and GoShimmer
- Updated goshimmerSeed module to the latest change
- Updated goshimmerGossipPort to the latest change
### Removed
- Removed watchdog log

---

## Version 0.8.6
### Changed
- Changed password setup for Bee dashboard
- Added pwd check for Hornet dashboard

---

## Version 0.8.5
### Added
- Added support for multiple domains
- Added SSH hardening setup
### Changed
- Changed proxy variables
- Changed proxy configuration menu
- Fixed general bugs

---

## Version 0.8.4
### Changed
- Fixed Goshimmer Seed
- Fixed issue with interal dashboard ports
- Fixed issue with internal api ports
- Fixed Bee pwd gen

---

## Version 0.8.3
- Hotfix for Bee installation

---

## Version 0.8.2
### Added
- Added converter for Hornet identityPrivateKey
### Changed
- Fixed missing port vars for Hornet and Bee
- Fixed beeVersion
- Set temporary Bee v0.2.0-rc1 as stable version

---

## Version 0.8.1
### Added
- Option to install Hornet and Bee
### Changed
- Fixed Hornet pruning database size
- Fixed Installer
- Fixed SWARM alias
- Fixed general bugs (includes bug fixes from 0.8.0)
### Removed
- Removed old code/files

---

## Version 0.7.9
- Hotfix for Bee

---

## Version 0.7.8
### Added
- Added comnet config to Bee
- Added new API to get versions
### Changed
- Moved alias to SWARM modules

dev-installer.sh
#!/bin/bash
clear
export NEWT_COLORS="
root=blue,blue
window=white
border=black
textbox=white
button=black,red
"
if [ "$1" = "-reinstall" ]; then
    if [ -d "/var/lib/swarm" ]; then
        sudo rm -rf /var/lib/swarm > /dev/null 2>&1
    fi
fi

if [ ! -f "/var/lib/swarm/swarm" ]; then
    if [ $(id -u) -ne 0 ]; then
        whiptail --title "SWARM" --msgbox "Please run SWARM installer with sudo or as root!" 8 65
        exit 0
    else
        if (whiptail --title "SWARM" --yesno "Do you want to install SWARM now?" 10 65); then
            whiptail --title "SWARM" --msgbox "SWARM will first make sure that your server is up to date..." 8 65
            sudo apt update
            sudo apt dist-upgrade -y
            sudo apt upgrade -y
            sudo apt autoremove -y
            {
                echo 0
                echo 10
                if [ ! -x "$(command -v git)" ]; then
                    sudo apt -qq install git -y > /dev/null 2>&1
                fi
                echo 30
                sudo git clone https://github.com/TangleBay/swarm.git /var/lib/swarm > /dev/null 2>&1
                ( cd /var/lib/swarm ; sudo git checkout develop > /dev/null 2>&1 )
                echo 50
                if [ -f "/var/lib/swarm/swarm" ]; then
                    source /var/lib/swarm/environment
                    sudo chmod +x $swarmHome/swarm $swarmPlugins/watchdog
                    echo 70
                    ( crontab -l | grep -v -F "$watchdogCronCmd" ; echo "$watchdogCronJob" ) | crontab - > /dev/null 2>&1
                    echo 100
                    clear
                fi
            } | whiptail --gauge "Please wait while SWARM is installed..." 8 65 0
            if [ -f "/var/lib/swarm/swarm" ]; then
                source /var/lib/swarm/environment
                source $swarmModules/swarmAlias
                clear
                if [ "$swarmAliasExists" = "true" ]; then
                    source $swarmHome/swarm
                fi
            else
                whiptail --title "SWARM" --msgbox "SWARM could not be successfully cloned from GitHub!" 8 65
            fi
        fi
    fi
else
    whiptail --title "SWARM" --msgbox "SWARM is already installed on your system!" 8 65
fi
exit 0

installer.sh
#!/bin/bash
clear
export NEWT_COLORS="
root=blue,blue
window=white
border=black
textbox=white
button=black,red
"
if [ "$1" = "-reinstall" ]; then
    if [ -d "/var/lib/swarm" ]; then
        sudo rm -rf /var/lib/swarm > /dev/null 2>&1
    fi
fi

if [ ! -f "/var/lib/swarm/swarm" ]; then
    if [ $(id -u) -ne 0 ]; then
        whiptail --title "SWARM" --msgbox "Please run SWARM installer with sudo or as root!" 8 65
        exit 0
    else
        if (whiptail --title "SWARM" --yesno "Do you want to install SWARM now?" 10 65); then
            whiptail --title "SWARM" --msgbox "SWARM will first make sure that your server is up to date..." 8 65
            sudo apt update
            sudo apt dist-upgrade -y
            sudo apt upgrade -y
            sudo apt autoremove -y
            {
                echo 0
                echo 10
                if [ ! -x "$(command -v git)" ]; then
                    sudo apt -qq install git -y > /dev/null 2>&1
                fi
                echo 30
                sudo git clone https://github.com/TangleBay/swarm.git /var/lib/swarm > /dev/null 2>&1
                if [ "$1" = "develop" ]; then
                    ( cd /var/lib/swarm ; sudo git checkout $1 > /dev/null 2>&1 )
                fi
                echo 50
                if [ -f "/var/lib/swarm/swarm" ]; then
                    source /var/lib/swarm/environment
                    sudo chmod +x $swarmHome/swarm $swarmPlugins/watchdog
                    echo 70
                    ( crontab -l | grep -v -F "$watchdogCronCmd" ; echo "$watchdogCronJob" ) | crontab - > /dev/null 2>&1
                    echo 100
                    clear
                fi
            } | whiptail --gauge "Please wait while SWARM is installed..." 8 65 0
            if [ -f "/var/lib/swarm/swarm" ]; then
                source /var/lib/swarm/environment
                source $swarmModules/swarmAlias
                clear
                if [ "$swarmAliasExists" = "true" ]; then
                    source $swarmHome/swarm
                fi
            else
                whiptail --title "SWARM" --msgbox "SWARM could not be successfully cloned from GitHub!" 8 65
            fi
        fi
    fi
else
    whiptail --title "SWARM" --msgbox "SWARM is already installed on your system!" 8 65
fi
exit 0